server.js: Add optional <profile> cmd line arg

kazarmy opened the pull request:
This pr adds a one-shot way via server.js to run Firefox Profiler with a local profile file. It can be used in tandem with a shell script like the following:

```sh
#!/bin/sh

PROFILEPATH=$(realpath "$1")
PWD=$(pwd)
cd ~/git/firefox-profiler; yarn start "$PROFILEPATH"; cd "$PWD"
```

I have seen and run [samply](https://github.com/mstange/samply) and I think that:

1. There really is no need to run more than 1 server.
2. It appears easier to customize and prototype in a dynamic language like JavaScript rather than a compiled language like Rust. For example, changing the opened browser here is just a matter of modifying `webpack.local-config.js`.

Note that I have nothing against samply, Rust, or the people working on them. I am quite happy with putting all this in a private fork or unmaintained public fork. If this pr was accepted, it would be somewhat harder to switch to a dev server from another package e.g. Vite, though probably not impossible.

Future work for this would include:

1. Adding support for opening a profile in a running server. This probably would not require working HMR.
2. Support for sources appear to be straightforward.

If you have read this far, thanks for the attention!

mstange commented:
I see, so the use case here is:

 1. You have a Chrome / node / `perf script` profile file, i.e. a type of profile which doesn't require symbolication.
 2. You want to load this profile into a locally-running development build of the profiler, because you are working on the profiler itself.

That seems like a fine use case to make easier.

FWIW, `samply load` is much more complicated because it supports symbolication for C++ / Rust code, and can be used to open unsymbolicated profiles from samply but also from Firefox, such as profiles generated by `MOZ_PROFILER_SHUTDOWN` or by manual calls to `Services.profiler.dumpProfileToFileAsync`. But if that functionality is not needed and samply is not already installed on the system anyway, then I agree that compiling samply just for this is quite overkill.

kazarmy commented:
That doesn't mean that samply can't be modified to support unsymbolicated profiles, but I did get the impression that samply was mainly an internal tool. I don't really expect a profiler user to have much trouble compiling samply though, even with the out-of-date rustc in Ubuntu 22.04.

kazarmy commented:
That doesn't mean that samply can't be modified to provide _source code_ for unsymbolicated profiles ... sorry about the confusion

kazarmy commented:
> ...  because you are working on the profiler itself.

For now at least, and because a development build is good enough for me right now. This pr might or might not be abuse of a dev server.

kazarmy commented:
> Just to clarify, what I meant to say is that I approve of the approach in this PR.

Thanks! Admittedly, it was hard to determine whether you were being sarcastic or not :sweat_smile:

> As for making the source view work for JS profiles in a secure way, I'm curious about your solution

It probably will just involve the user ensuring that the server is on 127.0.0.1 and not 0.0.0.0 or a public IP address.

kazarmy commented on server.js:
```diff
@@ -64,6 +74,36 @@ if (localConfigExists) {
+    if (prevSetupMiddlewares) {
+      middlewares = prevSetupMiddlewares(middlewares, devServer);
+    }
+    devServer.app.use(
+      '/profiles/',
+      express.static(path.resolve(path.dirname(argv.profile)))
+    );
```
This does expose every file under `path.dirname(argv.profile)`. It doesn't seem too serious for its intended use case but will probably need mitigation soon just in case someone changes the host away from localhost.

kazarmy replied:
Mitigation is done by filtering (cf97a59a425d3d03f05024d0d9b72fe7bf6f33c9)

julienw reviewed (commented):
Thanks for the patch!

I'm not a big fan of the approach but if that makes your life easier I'm not against it.

It would be good to update the contributing doc, for example [part 2 in this paragraph](https://github.com/firefox-devtools/profiler/blob/main/CONTRIBUTING.md#loading-in-profiles-for-development) could be the right location.

Looking at your initial description I'm thinking that because you're using a shell script already, maybe you could expose your local directory directly using either python's http.server or the local-web-server node module (that we already have installed in the profiler project as `node_modules/.bin/ws`) in background, then start our server in background too, and finally open the firefox window using the URL from this server.

But maybe another vision is that it's what our `server.js` should do for an easier development environment, and that's why, again, I'm not against this PR. So happy to move it forward!

kazarmy commented on server.js:
```diff
@@ -4,16 +4,26 @@
 );
 
+const argv = yargs(hideBin(process.argv))
+  .command('* [<profile>]', 'Open Firefox Profiler, on <profile> if included.')
```
On yargs 17.6.2, omitting the `*` doesn't work -- without it, yargs refuses to process a `profile` arg that is given on the command line. I did think about using `open` instead of the default command, but I don't see any other function for `server.js` other than to (eventually) open profiles.

Sorry about the `<>`, initially the `profile` arg was required, and I just slapped on the `[]` as it seemed the natural thing to do, without looking at the documentation. Thanks for the catch!

kazarmy commented on server.js:
```diff
@@ -64,6 +74,48 @@ if (localConfigExists) {
+      express.static(path.resolve(path.dirname(argv.profile)))
+    );
+    return middlewares;
+  };
```
I think that any reduction in dependency on the (possibly capricious) internal behavior of `webpack-dev-server` is a good thing, and I've integrated the code above. Changes include:

1. The port for the profile server is OS-allocated instead of fixed.
2. The host for the profile server is the same as for the webpack server.
3. CORS handling.
4. The `open` package is now an explicit dependency.

For consistency, the options in `serverConfig.open` are honored. Theoretically, the port for the webpack server can be OS-allocated as well but that should be in a different PR since it might be complicated.

kazarmy commented:
> It would be good to update the contributing doc, for example [part 2 in this paragraph](https://github.com/firefox-devtools/profiler/blob/main/CONTRIBUTING.md#loading-in-profiles-for-development) could be the right location.

I have added the shell script I'm using to `bin/` (after modifying it to use relative dirs) and updated both CONTRIBUTING.MD and loading-in-profiles.md (5d0a1ccc5a0cbed31da4d1962924c4bbc3940e59 c0c40bd365f642d2dab0030cfa6b31817b2e4c8d).

kazarmy commented on yarn.lock:
```diff
@@ -13151,6 +13160,19 @@ yargs@^17.3.1:
     y18n "^5.0.5"
     yargs-parser "^21.0.0"
 
+yargs@^17.6.2:
```
Okay 6723ea89ce42f173214fd8d30b531298a82da93a

kazarmy commented:
Okay I think that's it.

julienw commented:
Hey! Just a quick note that we've seen your update but just didn't have the chance to look at it during this week. It's still on our radar, thanks for your patience!

kazarmy commented:
No worries :) I see that you have a lot on your plate.

julienw commented on bin/fp.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+# This script is primarily for running Firefox Profiler on a saved profile.
```
Also what do you think about naming this file `launch-fp.sh` so that it's more obvious about what it does?

kazarmy replied:
> Can you please add a few examples about how to use it in the comments here?

Ok 974ad6d7738e5f73ac29e390f9ee9c31e6414b99

> Also what do you think about naming this file `launch-fp.sh` so that it's more obvious about what it does?

No strong opinions on this and I'm symlinking to it anyway so ok 974ad6d7738e5f73ac29e390f9ee9c31e6414b99

julienw commented on docs-developer/loading-in-profiles.md:
```diff
@@ -102,4 +102,4 @@ Firefox loads the profiles directly into the front-end through a WebChannel mech
 > `https://profiler.firefox.com/from-file/`
 
-When you're on [the home page](https://profiler.firefox.com) files can be loaded by either dragging over the profiler.firefox.com client, or using the file upload input.
+When you're on [the home page](https://profiler.firefox.com) files can be loaded by either dragging over the profiler.firefox.com client, or using the file upload input. On Linux, the provided [fp.sh](../bin/fp.sh) script can load files.
```
I feel like this addition should be in another paragraph though. Indeed we're really talking about the `from-file` datasource here.

What about moving this doc in the URL part above, below the part `Here is an example profile server written in Node.js:`:


> Note that if you have a copy of the project locally, you can simply add the file path as a parameter to `yarn start` and a server serving this file will be spawned for you. On Linux, the provided [launch-fp.sh](../bin/launch-fp.sh) script can do this for you as well, from any working directory.

kazarmy replied:
Ok lgtm with some minor editing (file -> profile, and removed 'simply' because alex complained) fe0bbab29c07ae25a948b2e412e7cdcdcbba6a4b fb1b549e23971d34a736658468ed837944db47ca

> Indeed we're really talking about the from-file datasource here.

The document looked like it was for users that are also developers rather than Firefox Profiler developers specifically.

julienw reviewed (commented):
Thanks for your patience!

I'm still not 100% convinced to have this as part of the server.js file. Indeed this would work only when the server isn't running yet; if it's already running, this would just fail. I feel like that a standalone nodejs script that would:
1. spawn a local server like you did already
2. open the profiler with a `from-url` URL properly filled in

would be more interesting.

This would have the following features:
1. support --release and --dev to decide whether to open profiler.firefox.com or localhost:4242
2. in the case of --release, we need a https server. So maybe a first implementation wouldn't support this case.
3. In the case of --dev, it would be good to check if the server runs already; if it doesn't run then in that case launch `yarn start` in the background first.

__that said__ I'm fine with landing your patch as it is with the small comments fixed :-)

kazarmy commented:
> Thanks for your patience!

Not sure whether I should have been pestering you lol.

> Indeed this would work only when the server isn't running yet; if it's already running, this would just fail.

> 3. In the case of --dev, it would be good to check if the server runs already; if it doesn't run then in that case launch yarn start in the background first.

The plan is to put down somewhere a port-specific flag file when the server starts and then to check that file on script start to see whether a server is running. Since the file might not be cleaned up if the server crashes, the user should be informed whenever a relevant flag file is detected.

> I feel like that a standalone nodejs script ...

This _might_ cause code duplication but it apparently makes it easier to switch dev servers if needed so sgtm.

julienw reviewed (approved):
Thx!

kazarmy commented:
> I feel like that a standalone nodejs script that would:
> 
> 1. spawn a local server like you did already
> 2. open the profiler with a `from-url` URL properly filled in
> 
> would be more interesting.

It seems a bit tricky to pull out the following code:

https://github.com/firefox-devtools/profiler/blob/69ed07f1336a9dc4f67f9ef87831ba16857e88a9/server.js#L94-L108

into that standalone nodejs script since that might require modifying the local `webpack.local-config.js` file somehow, but I'll probably figure out something by this week.

kazarmy commented:
"by this week" unfortunately has become "delayed indefinitely" but not losing hope that the above will eventually be resolved
